exchanges/okx/okx_websocket.go

```diff
@@ -0,0 +1,1858 @@
+	OkxChannelIndexCandle6Hutc  = indexCandlestick + OkxChannelCandle6Hutc
+
+	// Mark price candlesticks channel
+	OkxChannelMarkPriceCandle1Y     = markPrice + OkxChannelCandle1Y
+	OkxChannelMarkPriceCandle6M     = markPrice + OkxChannelCandle6M
+	OkxChannelMarkPriceCandle3M     = markPrice + OkxChannelCandle3M
+	OkxChannelMarkPriceCandle1M     = markPrice + OkxChannelCandle1M
+	OkxChannelMarkPriceCandle1W     = markPrice + OkxChannelCandle1W
+	OkxChannelMarkPriceCandle1D     = markPrice + OkxChannelCandle1D
+	OkxChannelMarkPriceCandle2D     = markPrice + OkxChannelCandle2D
+	OkxChannelMarkPriceCandle3D     = markPrice + OkxChannelCandle3D
+	OkxChannelMarkPriceCandle5D     = markPrice + OkxChannelCandle5D
+	OkxChannelMarkPriceCandle12H    = markPrice + OkxChannelCandle12H
+	OkxChannelMarkPriceCandle6H     = markPrice + OkxChannelCandle6H
+	OkxChannelMarkPriceCandle4H     = markPrice + OkxChannelCandle4H
+	OkxChannelMarkPriceCandle2H     = markPrice + OkxChannelCandle2H
+	OkxChannelMarkPriceCandle1H     = markPrice + OkxChannelCandle1H
+	OkxChannelMarkPriceCandle30m    = markPrice + OkxChannelCandle30m
+	OkxChannelMarkPriceCandle15m    = markPrice + OkxChannelCandle15m
+	OkxChannelMarkPriceCandle5m     = markPrice + OkxChannelCandle5m
+	OkxChannelMarkPriceCandle3m     = markPrice + OkxChannelCandle3m
+	OkxChannelMarkPriceCandle1m     = markPrice + OkxChannelCandle1m
+	OkxChannelMarkPriceCandle1Yutc  = markPrice + OkxChannelCandle1Yutc
+	OkxChannelMarkPriceCandle3Mutc  = markPrice + OkxChannelCandle3Mutc
+	OkxChannelMarkPriceCandle1Mutc  = markPrice + OkxChannelCandle1Mutc
+	OkxChannelMarkPriceCandle1Wutc  = markPrice + OkxChannelCandle1Wutc
+	OkxChannelMarkPriceCandle1Dutc  = markPrice + OkxChannelCandle1Dutc
+	OkxChannelMarkPriceCandle2Dutc  = markPrice + OkxChannelCandle2Dutc
+	OkxChannelMarkPriceCandle3Dutc  = markPrice + OkxChannelCandle3Dutc
+	OkxChannelMarkPriceCandle5Dutc  = markPrice + OkxChannelCandle5Dutc
+	OkxChannelMarkPriceCandle12Hutc = markPrice + OkxChannelCandle12Hutc
+	OkxChannelMarkPriceCandle6Hutc  = markPrice + OkxChannelCandle6Hutc
+)
+
+// WsConnect initiates a websocket connection
+func (ok *Okx) WsConnect() error {
+	if !ok.Websocket.IsEnabled() || !ok.IsEnabled() {
+		return errors.New(stream.WebsocketNotEnabled)
+	}
+	var dialer websocket.Dialer
+	dialer.ReadBufferSize = 8192
+	dialer.WriteBufferSize = 8192
+	var authDialer websocket.Dialer
+	authDialer.ReadBufferSize = 8192
+	authDialer.WriteBufferSize = 8192
+	err := ok.Websocket.Conn.Dial(&dialer, http.Header{})
+	if err != nil {
+		return err
+	}
+	if ok.Verbose {
+		log.Debugf(log.ExchangeSys, "Successful connection to %v\n",
+			ok.Websocket.GetWebsocketURL())
+	}
+	ok.Websocket.Conn.SetupPingHandler(stream.PingHandler{
+		UseGorillaHandler: true,
+		MessageType:       websocket.PingMessage,
+		Delay:             time.Second * 5,
+	})
+
+	ok.Websocket.SetCanUseAuthenticatedEndpoints(true)
+	if ok.IsWebsocketAuthenticationSupported() {
+		go func() {
```

Comment:
I used this because the output from the WsAuth() function can delay too much and create a blocking function call.
So, I made a login handler in the WsHandleData() method for the login response.